akka-stream/src/main/scala/akka/stream/scaladsl/RetryFlow.scala

```diff
@@ -0,0 +1,234 @@
+    BidiShape(externalIn, internalOut, internalIn, externalOut)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+
+    private var elementInProgress: OptionVal[In] = OptionVal.none
+    private var retryNo = 0
+    private var moreInternalDemand = false
+
+    setHandler(
+      externalIn,
+      new InHandler {
+        override def onPush(): Unit = {
+          val element = grab(externalIn)
+          elementInProgress = OptionVal.Some(element)
+          retryNo = 0
+          pushInternal(element)
+        }
+
+        override def onUpstreamFinish(): Unit =
+          if (elementInProgress.isEmpty) {
+            completeStage()
+          }
+      })
+
+    setHandler(
+      internalOut,
+      new OutHandler {
+
+        override def onPull(): Unit = {
+          if (elementInProgress.isEmpty) {
+            if (!hasBeenPulled(externalIn) && !isClosed(externalIn)) {
+              pull(externalIn)
+            }
+          } else moreInternalDemand = true
+        }
+
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          if (elementInProgress.isEmpty || !cause.isInstanceOf[NonFailureCancellation]) {
+            super.onDownstreamFinish(cause)
+          } else {
+            // emit elements before finishing
+            setKeepGoing(true)
+          }
+        }
+      })
+
+    setHandler(
+      internalIn,
+      new InHandler {
+        override def onPush(): Unit = {
+          val result = grab(internalIn)
+          elementInProgress match {
+            case OptionVal.None =>
+              failStage(
+                new IllegalStateException(
+                  s"inner flow emitted unexpected element $result; the flow must be one-in one-out"))
+            case OptionVal.Some((_, _)) if retryNo == maxRetries => pushExternal(result)
+            case OptionVal.Some(in) =>
+              decideRetry(in, result) match {
+                case None          => pushExternal(result)
+                case Some(element) => planRetry(element)
+              }
+          }
+        }
+      })
+
+    setHandler(externalOut, new OutHandler {
+      override def onPull(): Unit =
+        // external demand
+        if (!hasBeenPulled(internalIn)) pull(internalIn)
+    })
+
+    private def pushInternal(element: In): Unit = {
+      moreInternalDemand = false
+      push(internalOut, element)
+    }
+
+    private def pushExternal(result: Out): Unit = {
+      elementInProgress = OptionVal.none
+      push(externalOut, result)
+      if (isClosed(externalIn)) {
+        completeStage()
+      } else if (moreInternalDemand) {
+        pull(externalIn)
+      }
+    }
+
+    private def planRetry(element: In): Unit = {
+      val delay = BackoffSupervisor.calculateDelay(retryNo, minBackoff, maxBackoff, randomFactor)
+      elementInProgress = OptionVal.Some(element)
+      retryNo += 1
+      pull(internalIn)
+      scheduleOnce(element, delay)
```

Comment:
this looks strange. using the `element` as the timerKey. we don't know if the element has a good equals.
can the element be grabbed from `elementInProgress` in `onTimer` instead of secretly passing it as the timer key?

Reply:
Ok, I introduced a case object key and read the element from the option.